src/app/EventManagement.cpp

```diff
@@ -0,0 +1,903 @@
+                    // request
+                    SuccessOrExit(err);
+                    continue;
+                }
+                // we cannot copy event outright. We remember the
+                // current required space in mpAppData, we note the
+                // space requirements for the event in the current
+                // buffer and make that space in the next buffer.
+                eventBuffer->mAppData = reinterpret_cast<void *>(requiredSpace);
+                eventBuffer           = eventBuffer->mpNext;
+
+                // Sanity check: Die here on null event buffer.  If
+                // eventBuffer->mpNext were null, then the `EvictBuffer`
+                // in line 130 would have succeeded -- the event was
+                // already in the final buffer.
+                VerifyOrDie(eventBuffer != nullptr);
+
+                requiredSpace = ctx.mSpaceNeededForEvent;
+            }
+        }
+        else
+        {
+            if (eventBuffer == mpEventBuffer)
+                break;
+            eventBuffer   = eventBuffer->mpPrev;
+            requiredSpace = reinterpret_cast<size_t>(eventBuffer->mAppData);
+            err           = CHIP_NO_ERROR;
+        }
+    }
+
+    // On exit, configure the top-level s.t. it will always fail to evict an element
+    mpEventBuffer->mProcessEvictedElement = AlwaysFail;
+    mpEventBuffer->mAppData               = nullptr;
+
+exit:
+    return err;
+}
+
+CHIP_ERROR EventManagement::BlitEvent(EventLoadOutContext * apContext, EventLoggingDelegate * apDelegate,
+                                      const EventOptions * apOptions)
+{
+
+    CHIP_ERROR err       = CHIP_NO_ERROR;
+    TLVWriter checkpoint = apContext->mWriter;
+    chip::TLV::TLVType dataContainerType;
+    chip::app::EventDataElement::Builder eventDataElementBuilder;
+    chip::app::EventPath::Builder eventPathBuilder;
+
+    VerifyOrExit(apContext->mCurrentEventNumber >= apContext->mStartingEventNumber,
+                 /* no-op: don't write event, but advance current event Number */);
+
+    VerifyOrExit(apOptions != nullptr, err = CHIP_ERROR_INVALID_ARGUMENT);
+    VerifyOrExit(apOptions->mTimestamp.mType != Timestamp::Type::kInvalid, err = CHIP_ERROR_INVALID_ARGUMENT);
+
+    eventDataElementBuilder.Init(&(apContext->mWriter));
+    eventPathBuilder = eventDataElementBuilder.CreateEventPathBuilder();
+    SuccessOrExit(eventPathBuilder.GetError());
+
+    // TODO: Revisit NodeId since the the encoding spec and the IM seem to disagree on how this stuff works
+    eventPathBuilder.NodeId(apOptions->mpEventSchema->mNodeId)
+        .EndpointId(apOptions->mpEventSchema->mEndpointId)
+        .ClusterId(apOptions->mpEventSchema->mClusterId)
+        .EventId(apOptions->mpEventSchema->mEventId)
+        .EndOfEventPath();
+    SuccessOrExit(eventPathBuilder.GetError());
+
+    eventDataElementBuilder.PriorityLevel(static_cast<uint8_t>(apContext->mPriority));
```

Comment:
could we avoid a cast here and let PriorityLevel take an enum as input? Or is this stored somehow else?

Reply:
finally eventDataElementBuilder would store the value in tlv, which needs uint8_t casting